lib/synapse/haproxy.rb

```diff
@@ -1030,19 +1030,33 @@ def generate_backend_stanza(watcher, config)
       ]
     end
 
+    def haproxy_exec(command)
+      info = nil
+      s = UNIXSocket.new(@opts['socket_file_path'])
+      s.write(command)
+      info = s.read
+    rescue StandardError => e
+      log.warn "synapse: unhandled error reading stats socket: #{e.inspect}"
+      @restart_required = true
+      nil
+    ensure
+      unless s.nil?
+        begin
+          s.close
+        rescue StandardError => e
+          log.warn "synapse: unhandled error closing stats socket: #{e.inspect}"
+        end
+      end
+      info
+    end
+
     # tries to set active backends via haproxy's stats socket
     # because we can't add backends via the socket, we might still need to restart haproxy
     def update_backends(watchers)
       # first, get a list of existing servers for various backends
-      begin
-        s = UNIXSocket.new(@opts['socket_file_path'])
-        s.write("show stat\n")
-        info = s.read()
-      rescue StandardError => e
-        log.warn "synapse: unhandled error reading stats socket: #{e.inspect}"
-        @restart_required = true
-        return
-      end
+      info = haproxy_exec("show stat\n")
+
+      return if info.nil?
```

Comment:
i was confused here, because we appear to just exit without setting `@restart_required`. i suppose the expectation is that we only get `nil` output if there was an error, but this is why i was uncertain about the broad `rescue` block in the `haproxy_exec` function.